packages/hardhat-core/src/builtin-tasks/compile.ts

```diff
@@ -379,39 +381,105 @@ subtask(TASK_COMPILE_SOLIDITY_COMPILE_JOBS)
         return { artifactsEmittedPerJob: [] };
       }
 
-      // sort compilation jobs by compiler version
-      const sortedCompilationJobs = compilationJobs
-        .slice()
-        .sort((job1, job2) => {
-          return semver.compare(
-            job1.getSolcConfig().version,
-            job2.getSolcConfig().version
-          );
-        });
+      const { default: pMap } = await import("p-map");
+
+      log(`Compiling ${compilationJobs.length} jobs`);
+
+      const versionList: string[] = [];
+      for (const job of compilationJobs) {
+        const solcVersion = job.getSolcConfig().version;
+
+        if (!versionList.includes(solcVersion)) {
+          // versions older than 0.4.11 don't work with hardhat
+          // see issue https://github.com/nomiclabs/hardhat/issues/2004
+          if (
+            semver.lt(solcVersion, COMPILE_TASK_FIRST_SOLC_VERSION_SUPPORTED)
+          ) {
+            throw new HardhatError(
+              ERRORS.BUILTIN_TASKS.COMPILE_TASK_UNSUPPORTED_SOLC_VERSION,
+              {
+                version: solcVersion,
+                firstSupportedVersion:
+                  COMPILE_TASK_FIRST_SOLC_VERSION_SUPPORTED,
+              }
+            );
+          }
 
-      log(`Compiling ${sortedCompilationJobs.length} jobs`);
+          versionList.push(solcVersion);
+        }
+      }
 
-      const artifactsEmittedPerJob: ArtifactsEmittedPerJob = [];
-      for (let i = 0; i < sortedCompilationJobs.length; i++) {
-        const compilationJob = sortedCompilationJobs[i];
+      /**
+       * I'm not super proud of this, but we have to download the list before the compilers,
+       * otherwise the next downloads fail with a race condition.
+       *
+       * :)
+       */
+      const compilersCache = await getCompilersDir();
+      const downloader = new CompilerDownloader(compilersCache);
+      await downloader.downloadCompilersList();
```

Comment:
This always downloads the compiler list. This makes the compilation slower and prevents it from working offline.

Reply:
What if we remove this and don't download the compilers in parallel?

Reply:
I made this point to Franco as well, but I think if we're making changes to speed up compilation anyways, then ensuring it's as fast as currently possible (i.e. parallel downloads) should be the focus.

That being said, you're right that this would prohibit offline compilation. I would suggest instead that we just catch this one error and let the code continue. If they already have the necessary compilers downloaded and would otherwise be able to compile offline, then it'll just work, and if not then it'll fail in the next download attempt as usual.

Reply:
But I think downloading this here will, in 99% of the cases, make the compilation slower vs just trying to download the compilers serially and removing this logic. The reason is that most of the time the task that tries to download the compiler won't do anything, as they'll already be downloaded.

Reply:
What if we add a check here to download only if it doesn't exist already? Like the one that already exists in the Downloader class

Reply:
@alcuadrado found a couple of extra issue related to downloading the compilers in parallel.

The first one has to do with wasm compilers: if the native compilers are not working for some reason (for example, permissions issues), then the compilation jobs will try to download the wasm compilers. If the same version is downloaded in parallel, we get the original race condition again.

This can be fixed by also downloading the wasm list beforehand. And we need to check in both cases that they don't exist before downloading them for compilation to work offline, and to not add unnecessary download times to the compilation. At this point, and considering that downloading the compilers in parallel is better only in very specific cases (when you are compiling for the first time a project that uses multiple compilers), we think it makes more sense to do the simple thing and download them sequentially.

Btw, even if we would do all of that, there's still another problem: if someone uses a couple of new versions of the compiler, but the list already exists locally, then when the `GET_SOLC_BUILD` subtask is executed in parallel, the list will be downloaded in parallel, and then we get the same race condition again. This is a very edge case, but it's not impossible.

Reply:
> for example, permissions issues
The most common example would be running under a different architecture on linux, which got quiet common because of Docker and M1s.